### PR TITLE
fix: Use weak references to elements inside of interceptor objects

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Change list

This is a quick and poorly tested approach to address https://github.com/appium/appium/discussions/18860#discussioncomment-6731006
 
## Types of changes

- [ ] No changes in production code.
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java [hunk 1]
```diff
@@ -46,10 +48,36 @@ class AppiumElementLocator implements CacheableLocator {
...
+     * using @AndroidFindBy (-s), @iOSFindBy (-s) and @FindBy (-s) annotation
+     * sets
+     *
+     * @param searchContextReference     The context reference to use when finding the element
+     * @param by                         a By locator strategy
+     * @param shouldCache                is the flag that signalizes that elements which
+     *                                   are found once should be cached
+     * @param duration                   timeout parameter for the element to be found
+     */
+    public AppiumElementLocator(
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java:68]:
```suggestion
    AppiumElementLocator(
```
does it make sense to make this constructor package-private?

File: src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java [hunk 2]
```diff
@@ -61,15 +89,29 @@ class AppiumElementLocator implements CacheableLocator {
...
     }
 
+    private Optional<SearchContext> getSearchContext() {
+        if (searchContext != null) {
+            return Optional.of(searchContext);
+        }
+        if (searchContextReference != null) {
+            return Optional.ofNullable(searchContextReference.get());
+        }
+        return Optional.empty();
```
Comment (valfirst) [inline src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java:112]:
```suggestion
        return Optional.ofNullable(searchContextReference).map(WeakReference::get);
```

File: src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java (hunk 1 above)
Comment (valfirst) [reply src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java:68]: The same quesiton for all new API-s using `WeakReference`-s

**Review by valfirst — CHANGES_REQUESTED** [review]

File: src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java (hunk 1 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java]: changed

File: src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java (hunk 2 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/pagefactory/AppiumElementLocator.java]: changed

**Review by valfirst — APPROVED** [review]

**Pull request merged**